WordPress/src/main/java/org/wordpress/android/ui/themes/ThemeBrowserActivity.java

```diff
@@ -184,136 +189,166 @@ public void onSupportSelected(String themeId) {
-                                mThemeBrowserFragment.setRefreshing(false);
-                                if (mThemeBrowserFragment.getCurrentThemeTextView() != null) {
-                                    mThemeBrowserFragment.getCurrentThemeTextView().setText(mCurrentTheme.getName());
-                                    mThemeBrowserFragment.setCurrentThemeId(mCurrentTheme.getId());
-                                }
-                            }
-                            if (mThemeSearchFragment != null && mThemeSearchFragment.isVisible()) {
-                                mThemeSearchFragment.setRefreshing(false);
-                            }
-                        } catch (JSONException e) {
-                            AppLog.e(T.THEMES, e);
-                        }
-                    }
-                }, new ErrorListener() {
-                    @Override
-                    public void onErrorResponse(VolleyError response) {
-                        String themeId = ThemeTable.getCurrentThemeId(WordPress.wpDB.getDatabase(),
-                                String.valueOf(mSite.getSiteId()));
-                        mCurrentTheme = ThemeTable.getTheme(WordPress.wpDB.getDatabase(),
-                                String.valueOf(mSite.getSiteId()), themeId);
-                        if (mCurrentTheme != null && mThemeBrowserFragment != null) {
-                            if (mThemeBrowserFragment.getCurrentThemeTextView() != null) {
-                                mThemeBrowserFragment.getCurrentThemeTextView().setText(mCurrentTheme.getName());
-                                mThemeBrowserFragment.setCurrentThemeId(mCurrentTheme.getId());
-                            }
-                        }
-                    }
-                }
-        );
+        mDispatcher.dispatch(ThemeActionBuilder.newFetchCurrentThemeAction(mSite));
     }
 
-    protected Theme getCurrentTheme() {
-        return mCurrentTheme;
+    public void fetchWpComThemesIfSyncTimedOut(boolean force) {
+        long currentTime = System.currentTimeMillis();
+        if (force || currentTime - AppPrefs.getLastWpComThemeSync() > WP_COM_THEMES_SYNC_TIMEOUT) {
+            mDispatcher.dispatch(ThemeActionBuilder.newFetchWpComThemesAction());
+        }
+    }
+
+    public void fetchInstalledThemesIfJetpackSite() {
+        if (mSite.isJetpackConnected() && mSite.isUsingWpComRestApi() && !mIsFetchingInstalledPlugins) {
+            mDispatcher.dispatch(ThemeActionBuilder.newFetchInstalledThemesAction(mSite));
+            mIsFetchingInstalledPlugins = true;
+        }
+    }
+
+    public void activateTheme(String themeId) {
+        ThemeModel theme = new ThemeModel();
+        theme.setThemeId(themeId);
+        ThemeStore.ActivateThemePayload payload = new ThemeStore.ActivateThemePayload(mSite, theme);
+
+        if (mSite.isJetpackConnected() && mThemeStore.getInstalledThemeByThemeId(themeId) == null) {
```

Comment:
Shouldn't we check for `isJetpackConnected` and `isUsingWpComRestApi` at the start of this method? Don't we need that for the `newActivateThemeAction` dispatch?

Reply:
We do need the `isUsingWpComRestApi` check at the beginning now that you mention it, `isJetpackConnected` is not required though since .com sites are also supported.

`ThemeAction.ACTIVATE_THEMES` supports both .com and Jetpack sites. The difference with Jetpack sites is that the theme has to be installed first (which is why we have this conditional).